census_view.cpp

```diff
@@ -1897,6 +3148,203 @@ void CensusDVCView::UponPasteCensus(wxCommandEvent&)
+                << LMI_FLUSH
+                ;
+            }
+
+        for(int j = 0; j < lmi::ssize(headers); ++j)
+            {
+            if(exact_cast<tnr_date>(current_cell[headers[j]]))
+                {
+                static int const jdn_min = calendar_date::gregorian_epoch_jdn;
+                static int const jdn_max = calendar_date::last_yyyy_date_jdn;
+                static int const ymd_min = JdnToYmd(jdn_t(jdn_min)).value();
+                static int const ymd_max = JdnToYmd(jdn_t(jdn_max)).value();
+                int z = value_cast<int>(values[j]);
+                if(jdn_min <= z && z <= jdn_max)
+                    {
+                    ; // Do nothing: JDN is the default expectation.
+                    }
+                else if(ymd_min <= z && z <= ymd_max)
+                    {
+                    z = YmdToJdn(ymd_t(z)).value();
+                    values[j] = value_cast<std::string>(z);
+                    }
+                else
+                    {
+                    alarum()
+                        << "Invalid date " << values[j]
+                        << " for '" << headers[j] << "'"
+                        << " on line " << current_line << "."
+                        << LMI_FLUSH
+                        ;
+                    }
+                }
+            current_cell[headers[j]] = values[j];
+            }
+        current_cell.Reconcile();
+        current_cell.RealizeAllSequenceInput();
+        cells.push_back(current_cell);
+
+        status() << "Added cell number " << cells.size() << '.' << std::flush;
+        }
+
+    if(0 == current_line)
+        {
+        warning() << "No cells to paste." << LMI_FLUSH;
+        return;
+        }
+
+    if(!document().IsModified() && !document().GetDocumentSaved())
+        {
+        case_parms ().clear();
+        case_parms ().push_back(archetype);
+        class_parms().clear();
+        class_parms().push_back(archetype);
+        cell_parms ().swap(cells);
+        }
+    else if(configurable_settings::instance().census_paste_palimpsestically())
+        {
+        cell_parms().swap(cells);
+        // Cells that were copied from lmi have DOB and not IssueAge,
+        // so pasting them back in sets UseDOB. Force UseDOB for case
+        // and class defaults to prevent showing a UseDOB column with
+        // each cell set to "Yes".
+        for(auto& j : case_parms ()) {j["UseDOB"] = "Yes";}
+        for(auto& j : class_parms()) {j["UseDOB"] = "Yes";}
+        }
+    else
+        {
+        cell_parms().reserve(cell_parms().size() + cells.size());
+        std::back_insert_iterator<std::vector<Input>> iip(cell_parms());
+        std::copy(cells.begin(), cells.end(), iip);
+        }
+
+    auto const old_rows = grid_window_->GetNumberRows();
+    auto const old_cols = grid_window_->GetNumberCols();
+    auto const new_rows = grid_table_->GetRowsCount();
+    auto const new_cols = grid_table_->GetColsCount();
```

Comment:
There is something I don't understand here: rows did change already because the table uses `cell_parms().size()` and `cell_parms()` were updated above. But what about the columns? AFAICS they're only updated by `Update()`, which calls `update_visible_columns()`, which calls `set_visible_columns()` which actually affects `GetColsCount()` result. But `Update()` is called only later below.

Am I missing something or is this code wrong?

Or do we perhaps not care about columns at all here and should only compare rows? After all, the DVC version doesn't seem to do anything about them.

Reply:
> Or do we perhaps not care about columns at all here and should only compare rows?

Yes, you are right. The code works correctly because of calling `Update()`, but `old_cols` allways equals to `new_cols`.